tests/e2e/cluster_downgrade_test.go

```diff
@@ -40,30 +48,56 @@ func TestDowngradeUpgrade(t *testing.T) {
 	lastVersionStr := fmt.Sprintf("%d.%d", lastVersion.Major, lastVersion.Minor)
 
 	e2e.BeforeTest(t)
-	dataDirPath := t.TempDir()
 
-	epc := startEtcd(t, currentEtcdBinary, dataDirPath)
-	validateVersion(t, epc, version.Versions{Cluster: currentVersionStr, Server: currentVersionStr})
+	t.Logf("Create cluster with version %s", currentVersionStr)
+	epc := newCluster(t, currentEtcdBinary, clusterSize)
+	for i := 0; i < len(epc.Procs); i++ {
+		validateVersion(t, epc.Cfg, epc.Procs[i], version.Versions{Cluster: currentVersionStr, Server: currentVersionStr})
+	}
+	t.Logf("Cluster created")
 
+	t.Logf("etcdctl downgrade enable %s", lastVersion)
 	downgradeEnable(t, epc, lastVersion)
-	expectLog(t, epc, "The server is ready to downgrade")
-	validateVersion(t, epc, version.Versions{Cluster: lastVersionStr, Server: currentVersionStr})
+	t.Log("Write keys to ensure wal snapshot is created and cluster version set is snapshotted")
+	triggerSnaphsot(t, epc)
```

Comment:
maybe call it 'writeKeysToTriggerSnapshot'. 
The log-line could be put inside the function.

Reply:
Done